openwebstart/src/main/java/com/openwebstart/app/icon/ApplicationIconDownloadUtils.java

```diff
@@ -0,0 +1,95 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class ApplicationIconDownloadUtils {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ApplicationIconDownloadUtils.class);
+
+    public static BufferedImage downloadIcon(final Application application, final IconDimensions dimension) {
+        Assert.requireNonNull(application, "application");
+        Assert.requireNonNull(dimension, "dimension");
+        final BufferedImage loadedIcon =  downloadIcon(application.getJnlpFile(), IconKind.SHORTCUT, dimension)
+                .orElseGet(() -> downloadIcon(application.getJnlpFile(), IconKind.DEFAULT, dimension)
+                        .orElseGet(() -> downloadFavIcon(application.getJnlpFile())
+                                .orElseThrow(() -> new RuntimeException("No icon found for app"))));
+        return ImageUtils.resize(loadedIcon, dimension.getDimension(), dimension.getDimension());
+    }
+
+    private static Optional<BufferedImage> downloadIcon(final JNLPFile jnlpFile, final IconKind iconKind, final IconDimensions dimension) {
+        return getIconUrl(jnlpFile, iconKind, dimension)
+                .map(u -> {
+                    try {
+                        return downloadFromUrl(u);
+                    } catch (final IOException e) {
+                        LOG.warn("Error while downloading icon from url '" + u + "': " + e.getMessage());
+                        return null;
+                    }
+                });
+    }
+
+    private static Optional<BufferedImage> downloadFavIcon(final JNLPFile jnlpFile) {
+        Assert.requireNonNull(jnlpFile, "jnlpFile");
+
+        return Optional.ofNullable(jnlpFile.getNotNullProbableCodeBase())
+                .map(URL::getPath)
+                .map(FavIcon::possibleFavIconLocations)
+                .orElse(Collections.emptyList())
+                .stream()
+                .flatMap(l -> getFavIconUrlStreamForPath(jnlpFile, l))
+                .map(Result.of(ApplicationIconDownloadUtils::downloadFromUrl))
+                .filter(Result::isSuccessful)
+                .findFirst()
+                .map(Result::getResult);
+    }
+
+    private static Stream<? extends URL> getFavIconUrlStreamForPath(final JNLPFile jnlpFile, final String l) {
+        final List<URL> urls = new ArrayList<>();
+        try {
+            urls.add(FavIcon.favUrl("/", l, jnlpFile));
+        } catch (MalformedURLException ignore) {
+        }
+        try {
+            urls.add(FavIcon.favUrl("\\", l, jnlpFile));
+        } catch (MalformedURLException ignore) {
+        }
+        return urls.stream();
+    }
+
+    private static Optional<URL> getIconUrl(final JNLPFile jnlpFile, final IconKind iconKind, final IconDimensions dimension) {
+        Assert.requireNonNull(iconKind, "iconKind");
+        Assert.requireNonNull(dimension, "dimension");
+
+        return Optional.ofNullable(jnlpFile)
```

Comment:
can JNLPFile be null at this point??

Reply:
The other would be `Optional.ofNullable(file.getInformation())`. While we only c all this method the jnlpFile could only be null based on a bug. Since I need to create an Optional instance in any case I decided to go this way